### ISSUE TITLE
Reset task status if definition has changed

### ISSUE THREAD

**Issue by superlevure** [go-task/task#455]
Hi, 

First of all, thanks for the great tool! 

Currently if you use `sources` to watch for changes and update the task definition, the task will still be marked as `up-to-date` even if you completely changed the logic behind it. I think it would make sense to "reset" the task status whenever its definition has been updated.

**Comment by andreynering** [comment]
Hi @superlevure, and thanks for opening this issue!

I think this is an interesting proposal, but I'm not sure everyone would think this is the desired behavior. Hearing more opinions would be interesting.

---

Implementation note: we could use https://github.com/mitchellh/hashstructure to calculate the hash of the task and compute that in the checksum.

- This means it wouldn't work in timestamp mode
- This could have a side-effect of invalidating the checksum when upgrading the Task version if we make changes to the schema. Possible mitigation: Perhpaps considering only deps: and cmds: would be enough?

**Comment by postlund** [comment]
> I like the idea of including the task definition (in it's entirety) in the hash. It's likely very error prone to try to actually figure out how something may have changed and be selective.

I like that too, especially since that would mean that (environment) variables are taken into account for as well, which would otherwise go unnoticed.

**Comment by postlund** [comment]
> > A pessimistic approach, where any change to a Taskfile.yml would trigger a re-build, would be good enough for me.
> 
> I haven't had the chance to take advantage of the feature yet, so might well be missing something, but doesn't task already provide this capability?
> 
> ```yaml
> sources: 
>   - Taskfile.yml
> ```

Yes, should work, but it assumes you do not use any includes. Which I use a whole lot of.

**Comment by ghostsquad** [comment]
Even though you _can_ use sources (built in), it makes a lot of sense to make this default behavior.

**Comment by pd93** [comment]
I don't _fully_ agree with this change. I can definitely see the advantages of it in certain cases. However, there are definitely scenarios where this will cause problems.

There are a few different things being talked about here:

1. Hashing the entire `Taskfile.yml` and invalidating all task caches if it changes.
2. Hashing individual tasks and invalidating that task's hash if its definition changes.
3. Hashing a subset of task attributes and invalidating the tasks' hash if those attributes change.

---

The first option is not practical because of "setup tasks". For example:

```yaml
version: '3'

tasks:
  generate-one-time-certificate:
    status:
      - test -f certs/my-cert
    cmds:
      - gen-cert

  run-my-thing:
    deps: [generate-one-time-certificate]
```

I _absolutely_ do not want to regenerate this certificate if I change something unrelated in the Taskfile. But I don't think this is really the feature people are asking for.

---

The second option has problems too. If I change the task definition like this:

```diff
    version: '3'

    tasks:
      generate-one-time-certificate:
+++     desc: Generates a one-time certificate
        status:
          - test -f certs/my-cert
        cmds:
          - gen-cert
```

Do I _really_ want to rerun the Task? Probably not. The status check hasn't changed and the command to generate it hasn't changed either. Sometimes we only want to run commands if we _really_ need to and I would rather not run something when it should have run than run something when it shouldn't have. I don't think hashing the entire task definition makes sense.

---

> Perhaps considering only `deps:` and `cmds:` would be enough?

This makes a lot more sense as by changing _what_ the task does, you are invalidating what it did on the last run by definition. However, I'm still concerned that this change of default behaviour will catch someone out. (Do we consider this a breaking change?)

For me personally, I'm happy just to run `--force` when I update a definition, but the status checks are still passing - Especially now that we're going to have the [gentle force experiment](https://github.com/go-task/task/issues/1200).

**Comment by leaanthony** [comment]
This is currently an issue for us. We use a variable that is passed around the tasks to indicate if we are building for production (which affects most tasks in the chain). Duplicating all the tasks for the different types of builds is hard to maintain.
At the very least, I would expect the tasks to not be considered up to date if the variables passed to that definition are changed.
Thoughts?
